### PR TITLE
Frontend routing implementation (WIP)

### PR DESCRIPTION
**Pull request by Deraen**
Should the main package depend on these frontend modules?

### CODE REVIEW

File: modules/reitit-frontend/src/reitit/frontend.cljs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  [controller method]
+  (when-let [f (get controller method)]
+    (f (::params controller))))
+
+(defn- pad-same-length [a b]
+  (concat a (take (- (count b) (count a)) (repeat nil))))
+
+(defn apply-controllers
+  "Applies changes between current controllers and
+  those previously enabled. Resets controllers which
```
Comment (miikka) [inline modules/reitit-frontend/src/reitit/frontend.cljs]: [nit] `which` -> `whose`

File: modules/reitit-frontend/src/reitit/frontend.cljs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        match (or (reitit/match-by-path routes (.getPath uri))
+                  {:data {:name :not-found}})
+        q (query-params uri)
+        ;; Coerce if coercion enabled
+        c (if (:result match)
+            (coercion/coerce-request (:result match) {:query-params q
+                                                      :path-params (:params match)})
+            {:query q
+             :path (:param match)})
+        ;; Replace original params with coerced params
```
Comment (ikitommi) [inline modules/reitit-frontend/src/reitit/frontend.cljs]: Rest of reitit keeps the coerced paramaters in a separate `:parameters` map, while retaining the originals. Would be coherent to do so here too.

File: modules/reitit-frontend/src/reitit/frontend.cljs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                                      (pad-same-length new-controllers old-controllers))
+                                 (keep identity)
+                                 vec)]
+    (doseq [controller (map :old changed-controllers)]
+      (apply-controller controller :stop))
+    (doseq [controller (map :new changed-controllers)]
+      (apply-controller controller :start))
+    new-controllers))
+
+(defn hash-change [routes hash]
```
Comment (ikitommi) [inline modules/reitit-frontend/src/reitit/frontend.cljs]: routes => router

File: modules/reitit-frontend/src/reitit/frontend.cljs (hunk 2 above)
Comment (ikitommi) [inline modules/reitit-frontend/src/reitit/frontend.cljs]: Should there be a side-effecting function `init!` or such with a callback function - to be called when the route changes. For Reagent, one could pass in `(partial swap! app-state assoc ::view)`. the re-frame would use this too.

File: modules/reitit-re-frame/src/reitit/re-frame.cljs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      ::not-initialized)))
+
+(defn routed-view
+  "Renders the current view component.
+
+  Current view component is the :view from last :controllers entry with :view set."
+  []
+  (let [view @(re/subscribe [:routes/match])
+        ;; Select downmost controller that has view component
+        controller (first (filter :view (reverse (:controllers (:data view)))))]
```
Comment (ikitommi) [inline modules/reitit-re-frame/src/reitit/re-frame.cljs]: Today I think the `:view` should not be part of the Controller, it should be normal route data instead. With the current solution, the last view masks all previous ones. If there are multiple controllers in an endpoint, the order matters. e.g.

```clj
["/admin/users" {:controllers [users-controller admin-controller]}]
```

```clj
["/admin/users" {:controllers [admin-controller users-controller]}]
```

might render totally different uis. Instead:

```clj
["/admin/users" {:view users-view
                 :controllers [users-controller admin-controller]}]
```

**Comment by Deraen** [comment]
TODO: rename `reitit.frontend` to `reitit.browser`, update with latest code (e.g. query-string support) from app projects and separate controllers and other to other modules.

File: examples/frontend/src/frontend/core.cljs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+(defonce match (r/atom nil))
+
+(defn current-page []
+  [:div
+   (if @match
+     [:div [(:view (:data @match))]])
+   (pr-str @match)])
+
+(def routes
```
Comment (ikitommi) [inline examples/frontend/src/frontend/core.cljs]: looks complex. would something like this work:

```clj
(def routes
  (reitit.core/router
    [["/"
      {:name ::frontpage
       :view home-page}]
     ["/about"
      {:name ::about
       :view about-page}]]))
```

Comment (Deraen) [reply examples/frontend/src/frontend/core.cljs]: No.

**Pull request merged**